### PR TITLE
[Tuya Climate] Support both datapoint and pins for active state

### PR DESCRIPTION
**Pull request by zry98**
Add support for using both `active_state` and `heating_state_pin` and/or `cooling_state_pin`.

# What does this implement/fix?

This enables active state (action) reporting on some devices (e.g., BAC-002-ELW thermostat) using the active state datapoint only for active mode. Otherwise, if only set the datapoint, cannot get active state; if only set the state pins, cannot switch HVAC mode. With this change, when both are configured, the device will prefer active state from the pins, and mode from the datapoint.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3855

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
```yaml
climate:
  - platform: tuya
    id: hvac
    name: 'HVAC'
    supports_heat: true
    supports_cool: true
    switch_datapoint: 1
    active_state:
      datapoint: 102
      cooling_value: 0
      heating_value: 1
      fanonly_value: 2
    heating_state_pin: 13
    cooling_state_pin: 12
    target_temperature_datapoint: 2
    current_temperature_datapoint: 3
    temperature_multiplier: 0.5
    fan_mode:
      datapoint: 103
      auto_value: 0
      high_value: 1
      medium_value: 2
      low_value: 3
    preset:
      eco:
        datapoint: 5
        temperature: 25.0
    visual:
      min_temperature: 5.0
      max_temperature: 35.0
      temperature_step: 0.5
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by paravoid** [comment]
I may be misunderstanding the whole thing, but as far as I can tell, your device has a button to set the HVAC mode (heat/cool/dry), and that is represented in the 102 datapoint. That is a setting, rather than an active state report. Is that right? So, you can set the thermostat to "heating mode" by setting 102, but 102 will not report whether heating is active (e.g. if the set temperature is configured to a lower value than the current temperature). Is that all correct?

If so, then embedding this "mode configuration" into the active state reporting doesn't seem right to me. This seems orthogonal in nature. Perhaps we just need a `mode_datapoint` setting of some sort to capture this functionality?

(Drive-by contribution, I don't have any kind of special review/push rights)

**Comment by zry98** [comment]
> Perhaps we just need a `mode_datapoint` setting of some sort to capture this functionality?

@paravoid Yes, I wanted to do that at the beginning, but I don't want to change the current code too much.

> That is a setting, rather than an active state report.

> This seems orthogonal in nature.

Actually I think it's fine, since tuya MCU uses a datapoint both for setting and reporting at the same time, such as the [`fan_mode`](https://esphome.io/components/climate/tuya#:~:text=horizontal%20swing%20action.-,fan_mode,-(Optional)%3A%20Configuration) config field and the [`target_temperature_datapoint`](https://esphome.io/components/climate/tuya#:~:text=is%20not%20configured.-,target_temperature_datapoint,-(Required%2C%20int)). So, I don't think it would be necessary to add a new config field like `mode_datapoint` if they are using the same datapoint. It seems most Tuya thermostats are like this, e.g., https://github.com/esphome/feature-requests/issues/1541, https://github.com/esphome/issues/issues/5408.

**Comment by paravoid** [comment]
Is it reporting what mode (heating, cooling, etc.) you've configured it to through the buttons of the panel, or is it reporting whether it's _actively_ heating or cooling in this very moment?

In other words, if you set HVAC to "heating", have a current temperature of 24, and then set the target temperature to 20, will it report "idle", or "heating"?

**Comment by zry98** [comment]
An example:

HVAC off:
```plaintext
[16:19:48][D][climate:396]: 'HVAC' - Sending state:
[16:19:48][D][climate:399]:   Mode: OFF
[16:19:48][D][climate:401]:   Action: OFF
[16:19:48][D][climate:404]:   Fan Mode: HIGH
[16:19:48][D][climate:410]:   Preset: NONE
[16:19:48][D][climate:419]:   Current Temperature: 27.50°C
[16:19:48][D][climate:425]:   Target Temperature: 24.00°C
```

Set mode to cool from Home Assistant:
```plaintext
[16:20:01][D][tuya:305]: Datapoint 1 update to ON
[16:20:01][V][tuya.climate:012]: MCU reported switch is: ON
[16:20:01][D][climate:396]: 'HVAC' - Sending state:
[16:20:01][D][climate:399]:   Mode: COOL
[16:20:01][D][climate:401]:   Action: IDLE
[16:20:01][D][climate:404]:   Fan Mode: HIGH
[16:20:01][D][climate:410]:   Preset: NONE
[16:20:01][D][climate:419]:   Current Temperature: 28.00°C
[16:20:01][D][climate:425]:   Target Temperature: 24.00°C
```

After the relay closed:
```plaintext
[16:20:03][V][tuya.climate:127]: Cooling state pin changed to: ON
[16:20:03][D][climate:396]: 'HVAC' - Sending state:
[16:20:03][D][climate:399]:   Mode: COOL
[16:20:03][D][climate:401]:   Action: COOLING
[16:20:03][D][climate:404]:   Fan Mode: HIGH
[16:20:03][D][climate:410]:   Preset: NONE
[16:20:03][D][climate:419]:   Current Temperature: 28.00°C
[16:20:03][D][climate:425]:   Target Temperature: 24.00°C
```

Before this change, if `active_state` not set, it can't switch the HVAC mode (`Setting unknown datapoint 0`); if `cooling_state_pin` not set, it can't detect active state (`HEATING` or `COOLING` in this very moment). For example, if I set target temperature to something higher than current, it still shows `COOLING` as active state:
```plaintext
[16:32:48][D][climate:396]: 'HVAC' - Sending state:
[16:32:48][D][climate:399]:   Mode: COOL
[16:32:48][D][climate:401]:   Action: COOLING
[16:32:48][D][climate:404]:   Fan Mode: HIGH
[16:32:48][D][climate:410]:   Preset: NONE
[16:32:48][D][climate:419]:   Current Temperature: 28.00°C
[16:32:48][D][climate:425]:   Target Temperature: 30.00°C
```
instead of `IDLE`:
```
[16:35:16][D][climate:396]: 'HVAC' - Sending state:
[16:35:16][D][climate:399]:   Mode: COOL
[16:35:16][D][climate:401]:   Action: IDLE
[16:35:16][D][climate:404]:   Fan Mode: HIGH
[16:35:16][D][climate:410]:   Preset: NONE
[16:35:16][D][climate:419]:   Current Temperature: 28.00°C
[16:35:16][D][climate:425]:   Target Temperature: 30.00°C
```

**Comment by zry98** [comment]
> In other words, if you set HVAC to "heating", have a current temperature of 24, and then set the target temperature to 20, will it report "idle", or "heating"?

@paravoid Without this change, it will report "HEATING".

**Comment by ssieb** [comment]
I don't see why the active state dpid isn't required.  Can it ever not be there?  How would you control it without that?

**Comment by zry98** [comment]
@ssieb I guess it's due to that some devices with only one mode (maybe like some heating-only thermostats) don't have this datapoint, only the `switch_datapoint`.

File: esphome/components/tuya/climate/tuya_climate.cpp [hunk 1]
```diff
@@ -147,14 +143,18 @@ void TuyaClimate::control(const climate::ClimateCall &call) {
...
+        this->parent_->set_enum_datapoint_value(*this->active_state_id_, *this->active_state_heating_value_);
+      } else if (new_mode == climate::CLIMATE_MODE_COOL && this->supports_cool_) {
+        this->parent_->set_enum_datapoint_value(*this->active_state_id_, *this->active_state_cooling_value_);
+      } else if (new_mode == climate::CLIMATE_MODE_DRY && this->active_state_drying_value_.has_value()) {
+        this->parent_->set_enum_datapoint_value(*this->active_state_id_, *this->active_state_drying_value_);
+      } else if (new_mode == climate::CLIMATE_MODE_FAN_ONLY && this->active_state_fanonly_value_.has_value()) {
+        this->parent_->set_enum_datapoint_value(*this->active_state_id_, *this->active_state_fanonly_value_);
+      }
+    } else {
+      ESP_LOGE(TAG, "Active state (mode) datapoint not configured");
```
Comment (ssieb) [inline esphome/components/tuya/climate/tuya_climate.cpp]:
```suggestion
      ESP_LOGW(TAG, "Active state (mode) datapoint not configured");
```

**Review by ssieb — APPROVED** [review]

**Pull request merged**